src/bgw/job_stat.c

```diff
@@ -409,25 +414,16 @@ bgw_job_stat_insert_relation(Relation rel, int32 bgw_job_id, bool mark_start,
 void
 ts_bgw_job_stat_mark_start(int32 bgw_job_id)
 {
-	/* Use double-check locking */
+	Relation rel =
+		table_open(catalog_get_table_id(ts_catalog_get(), BGW_JOB_STAT), ShareRowExclusiveLock);
 	if (!bgw_job_stat_scan_job_id(bgw_job_id,
 								  bgw_job_stat_tuple_mark_start,
 								  NULL,
 								  NULL,
 								  RowExclusiveLock))
```

Comment:
```suggestion
		table_open(catalog_get_table_id(ts_catalog_get(), BGW_JOB_STAT), ShareRowExclusiveLock);
	if (!bgw_job_stat_scan_job_id(bgw_job_id,
								  bgw_job_stat_tuple_mark_start,
								  NULL,
								  NULL,
								  ShareRowExclusiveLock))
```
I don't quite understand what's going on here. Do we take 2 different locks on the same table at the same time, where one is a subset of the other?

Reply:
No, the `ShareRowExclusiveLock` is not a superset of `RowExclusiveLock` and the difference is that `ShareRowExclusiveLock` is self-exclusive while `RowExclusiveLock` is not. We need to pass a lock mode to `table_open` (this is asserted) and since the `RowExclusiveLock` is released after the scan, we use the `ShareRowExclusiveLock` to block other sessions from updating the lock before we have had the chance to insert a new row (if it is needed).